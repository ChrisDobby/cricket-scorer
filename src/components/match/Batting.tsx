import * as React from 'react';
import { Extras } from './Extras';
import { Batting as InningsBatting, BattingInnings, howOutDescription } from '../../domain';
import * as styles from './styles';

const battingContainer: React.CSSProperties = {
    width: '100%',
    padding: '20px',
    marginBottom: '20px',
};

const howOut = (innings?: BattingInnings): string => {
    if (!innings) { return ''; }

    return howOutDescription(innings.wicket);
};

interface InningsItemProps { innings?: BattingInnings; }
const Howout = (props: InningsItemProps) => (
    <div className="col-md-4">{howOut(props.innings)}</div>
);

interface CellItemProps {
    property: string;
    style: React.CSSProperties;
}
const CellItem = ({ property, style }: CellItemProps) => (
    <div className="col-md-1" style={style}>{property}</div>
);

export interface BattingProps { batting: InningsBatting; }

export const Batting = ({ batting }: BattingProps) => (
    <div className="col-xl-8 col-lg-12">
        <div style={battingContainer}>
            <div className="row" style={styles.headingRow}>
                <div className="col-md-7"><h6>Batsman</h6></div>
                <div className="col-md-1" style={styles.numberCell}><h6>Runs</h6></div>
                <div className="col-md-1" style={styles.numberCell}><h6>Balls</h6></div>
                <div className="col-md-1" style={styles.numberCell}><h6>Mins</h6></div>
                <div className="col-md-1" style={styles.numberCell}><h6>4s</h6></div>
                <div className="col-md-1" style={styles.numberCell}><h6>6s</h6></div>
            </div>    
            {batting.batters.map(batter =>
                <div className="row" style={styles.itemRow} key={batter.position}>
                    <div className="col-md-3">{batter.name}</div>
                    <Howout innings={batter.innings} />
                    <CellItem property={batter.innings ? batter.innings.runs.toString() : ''} style={styles.runsCell} />
                    <CellItem
                        property={batter.innings ? batter.innings.ballsFaced.toString() : ''}
                        style={styles.numberCell}
                    />
                    <CellItem property={batter.innings ? '0' : ''} style={styles.numberCell} />
                    <CellItem
                        property={batter.innings ? batter.innings.fours.toString() : ''}
                        style={styles.numberCell}
                    />
                    <CellItem
                        property={batter.innings ? batter.innings.sixes.toString() : ''}
                        style={styles.numberCell}
                    />
                </div>)}
            <Extras extras={batting.extras} />
        </div>
    </div>
);
