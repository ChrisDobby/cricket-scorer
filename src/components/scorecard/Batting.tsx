import * as React from 'react';
import Grid from '@material-ui/core/Grid';
import Typography from '@material-ui/core/Typography';
import Hidden from '@material-ui/core/Hidden';
import Divider from '@material-ui/core/Divider';
import { withStyles } from '@material-ui/core/styles';
import Extras from './Extras';
import {
    Batting as InningsBatting,
    BattingInnings,
    howOutDescription,
    unavailablDescription,
    Batter,
} from '../../domain';
import * as styles from './styles';
import * as globalStyles from '../styles';
import TextUpdateNotify from '../TextUpdateNotify';
import HeaderText from './HeaderText';

const calculateStrikeRate = (runs: number, balls: number) =>
    (runs === 0 || balls === 0 ? 0 : (runs / balls) * 100).toFixed(2);

const smallExtraDetailText = (innings?: BattingInnings): string =>
    innings
        ? `${innings.ballsFaced} balls, ${innings.fours} 4s, ${innings.sixes} 6s, SR ${calculateStrikeRate(
              innings.runs,
              innings.ballsFaced,
          )}`
        : '';

const howOut = (
    batter: Batter,
    getBowlerAtIndex: (index: number) => string,
    getFielderAtIndex: (index: number) => string,
    sameBowlerAndFielder: (bowlerIndex: number, fielderIndex: number) => boolean,
): string => {
    if (!batter.innings && typeof batter.unavailableReason === 'undefined') {
        return '';
    }

    return typeof batter.unavailableReason !== 'undefined'
        ? unavailablDescription(batter.unavailableReason)
        : howOutDescription(getBowlerAtIndex, getFielderAtIndex, sameBowlerAndFielder)(
              (batter.innings as BattingInnings).wicket,
          );
};

interface InningsItemProps {
    batter: Batter;
    getBowlerAtIndex: (index: number) => string;
    getFielderAtIndex: (index: number) => string;
    sameBowlerAndFielder: (bowlerIndex: number, fielderIndex: number) => boolean;
}
const Howout = (props: InningsItemProps) => (
    <Grid item xs={6} md={4}>
        <Typography variant="body2">
            <TextUpdateNotify
                text={howOut(props.batter, props.getBowlerAtIndex, props.getFielderAtIndex, props.sameBowlerAndFielder)}
            />
        </Typography>
    </Grid>
);

interface BattingProps {
    batting: InningsBatting;
    battingTeamPlayers: string[];
    score: number;
    wickets: number;
    totalOvers: string;
    classes: any;
    getBowlerAtIndex: (index: number) => string;
    getFielderAtIndex: (index: number) => string;
    sameBowlerAndFielder: (bowlerIndex: number, fielderIndex: number) => boolean;
    calculateMinutes: (innings: BattingInnings) => number;
}

const Batting = ({
    batting,
    battingTeamPlayers,
    score,
    wickets,
    totalOvers,
    classes,
    getBowlerAtIndex,
    getFielderAtIndex,
    sameBowlerAndFielder,
    calculateMinutes,
}: BattingProps) => (
    <Grid item lg={8} md={12} sm={12} xs={12}>
        <Grid container className={classes.header}>
            <Grid item xs={10} md={6}>
                <HeaderText>Batsman</HeaderText>
            </Grid>
            <Grid item xs={2} md={1}>
                <HeaderText style={styles.numberCell}>Runs</HeaderText>
            </Grid>
            <Hidden smDown>
                <Grid item md={1}>
                    <HeaderText style={styles.numberCell}>Balls</HeaderText>
                </Grid>
            </Hidden>
            <Hidden smDown>
                <Grid item xs={false} md={1}>
                    <HeaderText style={styles.numberCell}>Mins</HeaderText>
                </Grid>
            </Hidden>
            <Hidden smDown>
                <Grid item xs={false} md={1}>
                    <HeaderText style={styles.numberCell}>4s</HeaderText>
                </Grid>
            </Hidden>
            <Hidden smDown>
                <Grid item xs={false} md={1}>
                    <HeaderText style={styles.numberCell}>6s</HeaderText>
                </Grid>
            </Hidden>
            <Hidden smDown>
                <Grid item xs={false} md={1}>
                    <HeaderText style={styles.numberCell}>SR</HeaderText>
                </Grid>
            </Hidden>
        </Grid>
        {batting.batters.map((batter, idx) => (
            <React.Fragment key={idx}>
                <Grid container>
                    <Grid item xs={4} md={2}>
                        <Typography variant="body2">{battingTeamPlayers[batter.playerIndex]}</Typography>
                    </Grid>
                    <Howout
                        batter={batter}
                        getBowlerAtIndex={getBowlerAtIndex}
                        getFielderAtIndex={getFielderAtIndex}
                        sameBowlerAndFielder={sameBowlerAndFielder}
                    />
                    <Grid item xs={2} md={1}>
                        <Typography variant="body1" style={styles.runsCell}>
                            <TextUpdateNotify text={batter.innings ? batter.innings.runs.toString() : ''} />
                        </Typography>
                    </Grid>
                    <Hidden smDown>
                        <Grid item md={1}>
                            <Typography variant="body1" style={styles.numberCell}>
                                <TextUpdateNotify text={batter.innings ? batter.innings.ballsFaced.toString() : ''} />
                            </Typography>
                        </Grid>
                    </Hidden>
                    <Hidden smDown>
                        <Grid item md={1}>
                            <Typography variant="body1" style={styles.numberCell}>
                                <TextUpdateNotify text={batter.innings ? `${calculateMinutes(batter.innings)}` : ''} />
                            </Typography>
                        </Grid>
                    </Hidden>
                    <Hidden smDown>
                        <Grid item md={1}>
                            <Typography variant="body1" style={styles.numberCell}>
                                <TextUpdateNotify text={batter.innings ? batter.innings.fours.toString() : ''} />
                            </Typography>
                        </Grid>
                    </Hidden>
                    <Hidden smDown>
                        <Grid item md={1}>
                            <Typography variant="body1" style={styles.numberCell}>
                                <TextUpdateNotify text={batter.innings ? batter.innings.sixes.toString() : ''} />
                            </Typography>
                        </Grid>
                    </Hidden>
                    <Hidden smDown>
                        <Grid item md={1}>
                            <Typography variant="body1" style={styles.numberCell}>
                                <TextUpdateNotify
                                    text={
                                        batter.innings
                                            ? calculateStrikeRate(batter.innings.runs, batter.innings.ballsFaced)
                                            : ''
                                    }
                                />
                            </Typography>
                        </Grid>
                    </Hidden>
                    <Hidden mdUp>
                        <Grid item xs={4} />
                    </Hidden>
                    <Hidden mdUp>
                        <Grid item xs={8}>
                            <Typography variant="caption">
                                <TextUpdateNotify text={smallExtraDetailText(batter.innings)} />
                            </Typography>
                        </Grid>
                    </Hidden>
                </Grid>
                <Grid item xs={12}>
                    <Divider />
                </Grid>
            </React.Fragment>
        ))}
        <Extras extras={batting.extras} />
        <Grid container className={classes.header}>
            <Grid item xs={4} md={3}>
                <HeaderText>Total</HeaderText>
            </Grid>
            <Grid item md={4}>
                <HeaderText>
                    <TextUpdateNotify text={`for ${wickets} (${totalOvers} overs)`} />
                </HeaderText>
            </Grid>
            <Grid item xs={2} md={1}>
                <HeaderText style={styles.runsCell}>
                    <TextUpdateNotify text={score.toString()} />
                </HeaderText>
            </Grid>
        </Grid>
    </Grid>
);

export default withStyles(globalStyles.themedStyles)(Batting);
