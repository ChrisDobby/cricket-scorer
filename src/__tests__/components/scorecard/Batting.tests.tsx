import * as React from 'react';
import { render, cleanup } from 'react-testing-library';
import Batting from '../../../components/scorecard/Batting';

describe.skip('Batting', () => {
    beforeEach(cleanup);
    const batting = {
        extras: {
            byes: 0,
            legByes: 0,
            noBalls: 0,
            wides: 0,
            penaltyRuns: 0,
        },
    };

    it('should render correctly when batter has no innings', () => {
        const battingWithNoInnings = {
            ...batting,
            batters: [
                {
                    position: 1,
                    playerIndex: 0,
                },
            ],
        };

        const { container } = render(
            <Batting
                batting={battingWithNoInnings}
                battingTeamPlayers={['A Player']}
                getBowlerAtIndex={() => 'A player'}
                getFielderAtIndex={() => 'A player'}
                sameBowlerAndFielder={() => false}
                score={100}
                wickets={1}
                totalOvers={'20'}
                calculateMinutes={() => 1}
            />,
        );

        expect(container).toMatchSnapshot();
    });
});
