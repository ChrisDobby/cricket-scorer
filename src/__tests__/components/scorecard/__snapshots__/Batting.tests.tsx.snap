// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Batting should render correctly when batter has no innings 1`] = `
<div>
  <div
    class="MuiGrid-item-4 MuiGrid-grid-xs-12-43 MuiGrid-grid-sm-12-57 MuiGrid-grid-md-12-71 MuiGrid-grid-lg-8-81"
  >
    <div
      class="MuiGrid-container-3 Batting-header-1"
    >
      <div
        class="MuiGrid-item-4 MuiGrid-grid-xs-10-41 MuiGrid-grid-md-7-66"
      >
        <h6
          class="MuiTypography-root-100 MuiTypography-subtitle1-118 MuiTypography-colorInherit-129"
        >
          Batsman
        </h6>
      </div>
      <div
        class="MuiGrid-item-4 MuiGrid-grid-xs-2-33 MuiGrid-grid-md-1-60"
      >
        <h6
          class="MuiTypography-root-100 MuiTypography-subtitle1-118 MuiTypography-colorInherit-129"
          style="text-align: right;"
        >
          Runs
        </h6>
      </div>
      <div
        class="MuiGrid-item-4 MuiGrid-grid-md-1-60"
      >
        <h6
          class="MuiTypography-root-100 MuiTypography-subtitle1-118 MuiTypography-colorInherit-129"
          style="text-align: right;"
        >
          Balls
        </h6>
      </div>
      <div
        class="MuiGrid-item-4 MuiGrid-grid-md-1-60"
      >
        <h6
          class="MuiTypography-root-100 MuiTypography-subtitle1-118 MuiTypography-colorInherit-129"
          style="text-align: right;"
        >
          Mins
        </h6>
      </div>
      <div
        class="MuiGrid-item-4 MuiGrid-grid-md-1-60"
      >
        <h6
          class="MuiTypography-root-100 MuiTypography-subtitle1-118 MuiTypography-colorInherit-129"
          style="text-align: right;"
        >
          4s
        </h6>
      </div>
      <div
        class="MuiGrid-item-4 MuiGrid-grid-md-1-60"
      >
        <h6
          class="MuiTypography-root-100 MuiTypography-subtitle1-118 MuiTypography-colorInherit-129"
          style="text-align: right;"
        >
          6s
        </h6>
      </div>
    </div>
    <div
      class="MuiGrid-container-3"
    >
      <div
        class="MuiGrid-item-4 MuiGrid-grid-xs-4-35 MuiGrid-grid-md-3-62"
      >
        <p
          class="MuiTypography-root-100 MuiTypography-body2-108"
        >
          A Player
        </p>
      </div>
      <div
        class="MuiGrid-item-4 MuiGrid-grid-xs-6-37 MuiGrid-grid-md-4-63"
      >
        <p
          class="MuiTypography-root-100 MuiTypography-body2-108"
        >
          <span />
        </p>
      </div>
      <div
        class="MuiGrid-item-4 MuiGrid-grid-xs-2-33 MuiGrid-grid-md-1-60"
      >
        <p
          class="MuiTypography-root-100 MuiTypography-body1-109"
          style="text-align: right; font-weight: bold;"
        >
          <span />
        </p>
      </div>
      <div
        class="MuiGrid-item-4 MuiGrid-grid-md-1-60"
      >
        <p
          class="MuiTypography-root-100 MuiTypography-body1-109"
          style="text-align: right;"
        >
          <span />
        </p>
      </div>
      <div
        class="MuiGrid-item-4 MuiGrid-grid-md-1-60"
      >
        <p
          class="MuiTypography-root-100 MuiTypography-body1-109"
          style="text-align: right;"
        >
          <span />
        </p>
      </div>
      <div
        class="MuiGrid-item-4 MuiGrid-grid-md-1-60"
      >
        <p
          class="MuiTypography-root-100 MuiTypography-body1-109"
          style="text-align: right;"
        >
          <span />
        </p>
      </div>
      <div
        class="MuiGrid-item-4 MuiGrid-grid-md-1-60"
      >
        <p
          class="MuiTypography-root-100 MuiTypography-body1-109"
          style="text-align: right;"
        >
          <span />
        </p>
      </div>
    </div>
    <div
      class="MuiGrid-item-4 MuiGrid-grid-xs-12-43"
    >
      <hr
        class="MuiDivider-root-135"
      />
    </div>
    <div
      class="MuiGrid-container-3"
    >
      <div
        class="MuiGrid-item-4 MuiGrid-grid-xs-4-35 MuiGrid-grid-md-3-62"
      >
        <p
          class="MuiTypography-root-100 MuiTypography-body1-109"
        >
          Extras
        </p>
      </div>
      <div
        class="MuiGrid-item-4 MuiGrid-grid-xs-6-37 MuiGrid-grid-md-4-63"
      >
        <p
          class="MuiTypography-root-100 MuiTypography-body2-108"
        >
          <span>
            0nb 0wd 0b 0lb
          </span>
        </p>
      </div>
      <div
        class="MuiGrid-item-4 MuiGrid-grid-xs-2-33 MuiGrid-grid-md-1-60"
      >
        <p
          class="MuiTypography-root-100 MuiTypography-body1-109"
          style="text-align: right; font-weight: bold;"
        >
          <span>
            0
          </span>
        </p>
      </div>
    </div>
    <div
      class="MuiGrid-container-3 Batting-header-1"
    >
      <div
        class="MuiGrid-item-4 MuiGrid-grid-xs-4-35 MuiGrid-grid-md-3-62"
      >
        <h6
          class="MuiTypography-root-100 MuiTypography-subtitle1-118 MuiTypography-colorInherit-129"
        >
          Total
        </h6>
      </div>
      <div
        class="MuiGrid-item-4 MuiGrid-grid-md-4-63"
      >
        <h6
          class="MuiTypography-root-100 MuiTypography-subtitle1-118 MuiTypography-colorInherit-129"
        >
          <span>
            for 1 (20 overs)
          </span>
        </h6>
      </div>
      <div
        class="MuiGrid-item-4 MuiGrid-grid-xs-2-33 MuiGrid-grid-md-1-60"
      >
        <h6
          class="MuiTypography-root-100 MuiTypography-subtitle1-118 MuiTypography-colorInherit-129"
          style="text-align: right; font-weight: bold;"
        >
          <span>
            100
          </span>
        </h6>
      </div>
    </div>
  </div>
</div>
`;
